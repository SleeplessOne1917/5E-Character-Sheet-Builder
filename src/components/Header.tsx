import DiceSVG from "./svgs/DiceSVG";
import Link from "next/link";

const Header = (): JSX.Element => (
  <header className="flex justify-between bg-davey-jones-800 p-2 items-center fixed md:static w-full">
    <div className="flex justify-between ml-3 items-center">
      <Link href="/">
        <a>
          <DiceSVG
            backgroundClass="fill-heavens-down"
            foregroundClass="fill-red-wine-800"
            className="w-12 h-12 md:w-24 md:h-24"
          />
        </a>
      </Link>
      <h1 className="text-heavens-down text-lg md:text-3xl xl:text-5xl ml-3">
        D&D 5E Character Sheet Builder
      </h1>
    </div>
    <nav className="hidden md:block mr-20 w-1/6">
      <ul className="flex justify-around items-center list-none">
        <li className="relative">
          <Link href="/create">
            <a className="text-heavens-down text-lg lg:text-xl xl:text-2xl hover:text-cantaloupe-800 active:text-cantaloupe-800 focus:text-cantaloupe-800 hover:-translate-y-1 absolute -bottom-4">
              Create
            </a>
          </Link>
        </li>
        <li className="relative">
          <Link href="#">
            <a className="text-heavens-down text-lg lg:text-xl xl:text-2xl hover:text-cantaloupe-800 active:text-cantaloupe-800 focus:text-cantaloupe-800 hover:-translate-y-1 absolute -bottom-4 whitespace-nowrap">
              Log In
            </a>
          </Link>
        </li>
      </ul>
    </nav>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      className="h-12 w-12 stroke-heavens-down cursor-pointer md:hidden hover:stroke-cantaloupe-800 active:stroke-cantaloupe-800 focus:stroke-cantaloupe-800"
      fill="none"
      viewBox="0 0 24 24"
      stroke="currentColor"
      strokeWidth={2}
    >
      <path
        strokeLinecap="round"
        strokeLinejoin="round"
        d="M4 6h16M4 12h16M4 18h16"
      />
    </svg>
  </header>
);

export default Header;
