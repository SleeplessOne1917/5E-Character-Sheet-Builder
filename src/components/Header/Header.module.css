.header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: var(--davy);
  padding: 0.5rem;
  z-index: 10;
  width: 100%;
  font-family: var(--font-fantasy);
}

.brand-container {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-left: 0.75rem;
}

.logo-background {
  fill: var(--heaven);
}

.logo-foreground {
  fill: var(--wine-red);
}

.logo {
  width: 3rem;
  height: 3rem;
}

.brand-text {
  color: var(--heaven);
  margin-left: 0.75rem;
  font-size: 1.125rem;
  line-height: 1.75rem;
}

.navigation {
  display: hidden;
  margin-right: 5rem;
  width: 16.666667%;
}

.navigation-list {
  display: flex;
  justify-content: space-around;
  align-items: center;
  list-style: none;
  position: relative;
}

.navigation-link {
  color: var(--heaven);
  font-size: 1.25rem;
  line-height: 1.75rem;
  position: absolute;
  bottom: -1rem;
  white-space: nowrap;
}

.navigation-link:hover,
.navigation-link:active,
.navigation:focus {
  color: var(--highlight);
}

.navigation:hover,
.navigation:active {
  transform: translateY(-0.25rem);
}

.menu-icon {
  height: 3rem;
  width: 3rem;
  stroke: var(--heaven);
  color: var(--heaven);
  cursor: pointer;
}

.menu-icon:hover,
.menu-icon:active,
.menu-icon:focus {
  stroke: var(--highlight);
  color: var(--highlight);
}

@media screen and (min-width: var(--medium)) {
  .logo {
    height: 6rem;
    width: 6rem;
  }

  .brand-text {
    font-size: 1.875rem;
    line-height: 2.25rem;
  }

  .navigation {
    display: block;
  }

  .menu-icon {
    display: none;
  }
}

@media screen and (min-width: var(--large)) {
  .navigation-link {
    font-size: 1.5rem;
    line-height: 2rem;
  }
}

@media screen and (min-width: var(--extra-large)) {
  .brand-text {
    font-size: 3rem;
    line-height: 1;
  }

  .navigation-link {
    font-size: 1.875rem;
    line-height: 2.25rem;
  }
}
