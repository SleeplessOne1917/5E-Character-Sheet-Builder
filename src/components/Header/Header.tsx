import Link from 'next/link';
import { MenuIcon } from '@heroicons/react/solid';
import { MouseEventHandler } from 'react';
import classes from './Header.module.css';
import { useAppSelector } from '../../hooks/reduxHooks';
import useLogout from '../../hooks/useLogout';

export type HeaderProps = {
	onMenuIconClick: MouseEventHandler<SVGSVGElement>;
	onLogoIconClick: MouseEventHandler<HTMLAnchorElement>;
};

const Header = ({
	onMenuIconClick,
	onLogoIconClick
}: HeaderProps): JSX.Element => {
	const viewer = useAppSelector(state => state.viewer);
	const logout = useLogout();

	return (
		<header className={classes.header} aria-label="Header" role="banner">
			<div className={classes['brand-container']}>
				<Link href="/">
					<a onClick={onLogoIconClick} aria-label="Home" data-testid="home">
						<svg className={classes.logo}>
							<use xlinkHref="/Icons.svg#logo" />
						</svg>
					</a>
				</Link>
				<div className={classes['brand-text']}>
					D&D 5E Character Sheet Builder
				</div>
			</div>
			<nav className={classes.navigation}>
				<ul className={classes['navigation-list']}>
					<li className={classes['navigation-list-item']}>
						<Link href="/create">
							<a className={classes['navigation-link']}>Create</a>
						</Link>
					</li>
					{!viewer && (
						<>
							<li className={classes['navigation-list-item']}>
								<Link href="/log-in">
									<a className={classes['navigation-link']}>Log In</a>
								</Link>
							</li>
							<li className={classes['navigation-list-item']}>
								<Link href="/sign-up">
									<a className={classes['navigation-link']}>Sign Up</a>
								</Link>
							</li>
						</>
					)}
					{viewer && (
						<li className={classes['navigation-list-item']}>
							<Link href="#">
								<a className={classes['navigation-link']} onClick={logout}>
									Log Out
								</a>
							</Link>
						</li>
					)}
				</ul>
			</nav>
			<MenuIcon
				className={classes['menu-icon']}
				onClick={onMenuIconClick}
				data-testid="menu"
			/>
		</header>
	);
};

export default Header;
