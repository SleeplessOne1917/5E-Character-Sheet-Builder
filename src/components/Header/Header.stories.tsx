import { ComponentMeta, ComponentStory } from '@storybook/react';

import Header from './Header';
import { Provider } from 'react-redux';
import { getTestStore } from '../../redux/store';
import loggedInViewerMock from '../../mock/loggedInViewerMock';

export default {
	title: 'Components/Header',
	component: Header,
	argTypes: {
		onLogoIconClick: { type: 'function' },
		onMenuIconClick: { type: 'function' }
	}
} as ComponentMeta<typeof Header>;

const Template: ComponentStory<typeof Header> = args => <Header {...args} />;

export const LoggedOut = Template.bind({});
LoggedOut.decorators = [
	Story => (
		<Provider store={getTestStore()}>
			<Story />
		</Provider>
	)
];

export const LoggedIn = Template.bind({});
LoggedIn.decorators = [
	Story => (
		<Provider store={getTestStore({ viewer: loggedInViewerMock })}>
			<Story />
		</Provider>
	)
];
