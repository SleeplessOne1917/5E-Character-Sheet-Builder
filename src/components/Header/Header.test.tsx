import '@testing-library/jest-dom';

import * as Router from 'next/router';
import * as stories from './Header.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';

const { LoggedIn, LoggedOut } = composeStories(stories);

beforeAll(() => {
	const useRouter = jest.spyOn(Router, 'useRouter');
	const router: Router.NextRouter = {
		push: () => Promise.resolve(true),
		prefetch: () => Promise.resolve(),
		route: '',
		query: {},
		pathname: '',
		asPath: '',
		basePath: '',
		isLocaleDomain: true,
		reload: jest.fn(),
		replace: jest.fn(),
		back: jest.fn(),
		beforePopState: jest.fn(),
		events: { emit: jest.fn(), on: jest.fn(), off: jest.fn() },
		isFallback: false,
		isPreview: false,
		isReady: true
	};
	useRouter.mockReturnValue(router);
});

describe('renders correctly', () => {
	it('when logged out', () => {
		render(<LoggedOut />);

		expect(screen.getByRole('banner')).toMatchSnapshot();
	});

	it('when logged in', () => {
		render(<LoggedIn />);

		expect(screen.getByRole('banner')).toMatchSnapshot();
	});
});

describe('has expected links', () => {
	it('when logged in', () => {
		render(<LoggedIn />);

		expect(screen.getByText(/Create/i)).toBeInTheDocument();
		expect(screen.queryByText(/Log In/i)).not.toBeInTheDocument();
		expect(screen.queryByText(/Sign Up/i)).not.toBeInTheDocument();
		expect(screen.getByText(/Log Out/i)).toBeInTheDocument();
	});

	it('when logged out', () => {
		render(<LoggedOut />);

		expect(screen.getByText(/Create/i)).toBeInTheDocument();
		expect(screen.getByText(/Log In/i)).toBeInTheDocument();
		expect(screen.getByText(/Sign Up/i)).toBeInTheDocument();
		expect(screen.queryByText(/Log Out/i)).not.toBeInTheDocument();
	});
});

it('calls onMenuIconClick when menu icon is clicked', async () => {
	const mockOnMenuIconClick = jest.fn();

	render(<LoggedOut onMenuIconClick={mockOnMenuIconClick} />);
	await userEvent.click(screen.getByTestId('menu'));

	expect(mockOnMenuIconClick).toHaveBeenCalled();
});
