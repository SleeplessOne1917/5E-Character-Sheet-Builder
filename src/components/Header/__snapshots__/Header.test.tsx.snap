// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when logged in 1`] = `
<header
  aria-label="Header"
  class="header"
  role="banner"
>
  <div
    class="brand-container"
  >
    <a
      aria-label="Home"
      data-testid="home"
      href="/"
    >
      <svg
        class="logo"
      >
        <use
          xlink:href="/Icons.svg#logo"
        />
      </svg>
    </a>
    <div
      class="brand-text"
    >
      D&D 5E Character Sheet Builder
    </div>
  </div>
  <nav
    class="navigation"
  >
    <ul
      class="navigation-list"
    >
      <li
        class="navigation-list-item"
      >
        <a
          class="navigation-link"
          href="/create"
        >
          Create
        </a>
      </li>
      <li
        class="navigation-list-item"
      >
        <a
          class="navigation-link"
          href="/my-stuff"
        >
          My Stuff
        </a>
      </li>
      <li
        class="navigation-list-item"
      >
        <a
          class="navigation-link"
          href="/account"
        >
          Account
        </a>
      </li>
      <li
        class="navigation-list-item"
      >
        <a
          class="navigation-link"
          href="#"
        >
          Log Out
        </a>
      </li>
    </ul>
  </nav>
  <svg
    aria-hidden="true"
    class="menu-icon"
    data-testid="menu"
    fill="currentColor"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clip-rule="evenodd"
      d="M3 6.75A.75.75 0 013.75 6h16.5a.75.75 0 010 1.5H3.75A.75.75 0 013 6.75zM3 12a.75.75 0 01.75-.75h16.5a.75.75 0 010 1.5H3.75A.75.75 0 013 12zm0 5.25a.75.75 0 01.75-.75h16.5a.75.75 0 010 1.5H3.75a.75.75 0 01-.75-.75z"
      fill-rule="evenodd"
    />
  </svg>
</header>
`;

exports[`renders correctly when logged out 1`] = `
<header
  aria-label="Header"
  class="header"
  role="banner"
>
  <div
    class="brand-container"
  >
    <a
      aria-label="Home"
      data-testid="home"
      href="/"
    >
      <svg
        class="logo"
      >
        <use
          xlink:href="/Icons.svg#logo"
        />
      </svg>
    </a>
    <div
      class="brand-text"
    >
      D&D 5E Character Sheet Builder
    </div>
  </div>
  <nav
    class="navigation"
  >
    <ul
      class="navigation-list"
    >
      <li
        class="navigation-list-item"
      >
        <a
          class="navigation-link"
          href="/create"
        >
          Create
        </a>
      </li>
      <li
        class="navigation-list-item"
      >
        <a
          class="navigation-link"
          href="/log-in"
        >
          Log In
        </a>
      </li>
      <li
        class="navigation-list-item"
      >
        <a
          class="navigation-link"
          href="/sign-up"
        >
          Sign Up
        </a>
      </li>
    </ul>
  </nav>
  <svg
    aria-hidden="true"
    class="menu-icon"
    data-testid="menu"
    fill="currentColor"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clip-rule="evenodd"
      d="M3 6.75A.75.75 0 013.75 6h16.5a.75.75 0 010 1.5H3.75A.75.75 0 013 6.75zM3 12a.75.75 0 01.75-.75h16.5a.75.75 0 010 1.5H3.75A.75.75 0 013 12zm0 5.25a.75.75 0 01.75-.75h16.5a.75.75 0 010 1.5H3.75a.75.75 0 01-.75-.75z"
      fill-rule="evenodd"
    />
  </svg>
</header>
`;
